src/main/java/fi/aalto/cs/apluscourses/ui/base/TreeView.java

```diff
@@ -42,24 +57,38 @@ public TreeView() {
    */
   public void setViewModel(@Nullable BaseTreeViewModel<?> viewModel) {
     if (viewModel != null) {
-      DefaultTreeModel treeModel = new DefaultTreeModel(createNode(viewModel));
-      viewModel.filtered.addListener(treeModel, DefaultTreeModel::reload);
-      setModel(treeModel);
+      synchronized (viewModelLock) {
+        unregisterViewModel();
+        this.viewModel = viewModel;
+        registerViewModel();
+      }
+      update();
     }
     selectedItem = null;
   }
 
-  @NotNull
-  private static DefaultMutableTreeNode createNode(@NotNull Tree tree) {
-    List<? extends Tree> children = tree.getChildren();
-    OurTreeNode node = new OurTreeNode(tree);
-    if (tree instanceof Filterable) {
-      ((Filterable) tree).addVisibilityListener(node);
+  private void unregisterViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.removeCallback(this);
+      }
     }
-    for (Tree subtree : children) {
-      node.add(createNode(subtree));
+  }
+
+  private void registerViewModel() {
+    synchronized (viewModelLock) {
+      if (viewModel != null) {
+        viewModel.filtered.addListener(this, TreeView::update, SwingUtilities::invokeLater);
```

Comment:
[IntelliJ doc](https://github.com/JetBrains/intellij-community/blob/cf665c226927d79da572338d860b5ff7ff5aa812/platform/core-api/src/com/intellij/openapi/application/ModalityState.java#L44)
> SwingUtilities#invokeLater(Runnable) [...] may be used in the purely UI-related code, but not with anything that deals with PSI or VFS.

I'd say `SwingUtilities::invokeLater` is the correct choice here. This class is independent of IntelliJ framework; introducing `Application` reference here would make it dependent on the framework (and kinda justify itself). `update` method is not expected to call any framework-critical code, so there's no danger in that.

As our project grows, the modularity of the code becomes more and more important factor to keep up maintainability. Being strict with component dependencies and interfaces is the best approach for that. So even though calling `Application::invokeLater` here wouldn't hurt much on modularity itself, it would be yet another step on the slippery slope towards more component-model-breaking cross-references that ultimately turn the code into unmanageable spaghetti.